File: 백준/Gold/7569. 토마토/토마토.py
import sys
input = sys.stdin.readline
from collections import deque

M, N, H = map(int, input().split()) # 가로(열), 세로(행), 높이
tomatos = [[list(map(int, input().split())) for _ in range(N)] for _ in range(H)] # 3차원 배열 입력 받기

Q = deque()

# 익은 토마토 찾기
for h in range(H):
    for r in range(N):
        col = [idx for idx, value in enumerate(tomatos[h][r]) if value == 1]
        for c in col:
            Q.append((h, r, c))

days = 0
# bfs
while Q:
    curH, curR, curC = Q.popleft()
    for h, r, c in [(0, 1, 0), (0, -1, 0), (0, 0, 1), (0, 0, -1), (1, 0, 0), (-1, 0, 0)]: # 상하우좌앞뒤
        newH = curH + h
        newR = curR + r
        newC = curC + c

        if 0 <= newH <H and 0 <= newR < N and 0<= newC < M and tomatos[newH][newR][newC] == 0:
            # 새 좌표가 범위에 벗어나지 않으면서, 새 좌표의 토마토가 안 익었다면(방문을 안 했다면)
            tomatos[newH][newR][newC] = 1 + tomatos[curH][curR][curC]
            days = max(days, tomatos[newH][newR][newC])
            Q.append((newH, newR, newC))

flag = 0 # 숙성 못 하는 게 존재하는지 확인하는 용도의 플래그
for h in range(H):
    for r in range(N):
        if 0 in tomatos[h][r]:
            flag = 1
    if flag:
        break

# 숙성 못 하는 게 하나라도 있으면 -1 출력, 아니면 필요한 날짜 수 출력
if flag:
    print(-1)
else:
    if days: # 'days == 0이면' === '이미 다 숙성된 상태면'
        print(days - 1)
    else:
        print(days)

File: 백준/Gold/7569. 토마토/토마토.js
const fs = require('fs')
const input = fs.readFileSync(process.platform == 'linux' ? '/dev/stdin' : './7569input.txt').toString().trim().split('\n').map((el)=>el.split(' ').map(Number))

const [M, N, H] = input.shift() // 열, 행, 높이

const tomatos = [] // 토마토를 담을 3차원 배열
let array = [] // 임시 배열
let i = 0
// tomatos 배열 채우기
input.map(el => {
  array.push(el)
  i += 1
  if (i === N) {
    i = 0
    tomatos.push(array)
    array = []
  }
})

// deque 직접 구현하기
class Queue {
  constructor() {
    this.queue = []
    this.head = 0
  }

  enqueue(value) {
    this.queue.push(value)
  }

  dequeue() {
    return this.queue[this.head++]
  }
  isEmpty() {
    return this.queue.length === this.head
  }
}

const Q = new Queue()

// 익은 토마토 좌표 찾아서 Q에 담기
tomatos.map((tomatoBox, hIdx)=>{
  tomatoBox.map((tomatoBoxRow, rIdx)=> {
    let cIdx = tomatoBoxRow.indexOf(1)
    while (cIdx != -1) {
      Q.enqueue([hIdx, rIdx, cIdx])
      cIdx = tomatoBoxRow.indexOf(1, cIdx+1)
    }
  })
})

let days = 0
// bfs
const addArray = [[0, 1, 0], [0, -1, 0], [0, 0, 1], [0, 0, -1], [1, 0, 0], [-1, 0, 0]]
while (!Q.isEmpty()) {
  const [curH, curR, curC] = Q.dequeue()
  addArray.map(add => {
    const nextH = curH + add[0]
    const nextR = curR + add[1]
    const nextC = curC + add[2]
    if (nextH >= 0 && nextH < H && nextR >= 0 && nextR < N && nextC >= 0 && nextC < M && tomatos[nextH][nextR][nextC] === 0) {
      // 새 좌표가 범위에 벗어나지 않았으면서, 아직 안 익은 토마토가 있으면
      tomatos[nextH][nextR][nextC] = tomatos[curH][curR][curC] + 1
      Q.enqueue([nextH, nextR, nextC])
      days = Math.max(days, tomatos[nextH][nextR][nextC])
    }
  })
}

// 안 익은 토마토가 존재한다면 -1 출력하고, 아니면 걸린 시간 출력
let flag = -1
for (let h = 0; h < H; h++) {
  for (let r = 0; r < N; r++) {
    flag = tomatos[h][r].indexOf(0)
    if (flag !== -1) {
      break
    }
  }
  if (flag !== -1) {
    break
  }
}

if (flag !== -1) {
  console.log(-1)
} else {
  if (days === 0) {
    console.log(days)
  } else {
    console.log(days - 1)
  }
}